GH2400: Add globber pattern support to the #load directive

gitfool opened the pull request:
Fixes #2400. @patriksvensson almost there; a couple of the existing unit tests are failing though:

- [LoadDirectiveProcessorTests.cs#L168](https://github.com/gitfool/cake/blob/2ed89ca5be268818a58c3b4840a199aba0bd6af1/src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs#L168)
- [LoadDirectiveProcessorTests.cs#L206](https://github.com/gitfool/cake/blob/2ed89ca5be268818a58c3b4840a199aba0bd6af1/src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs#L206)

In both cases the globber is processing absolute paths that don't have any globber patterns in them and returning no files. Any ideas how to handle that?

gitfool commented:
@devlead @patriksvensson I had some time to dig into this; found and fixed the issue with the globber not working as expected with unix root paths. Also fixed some Windows tests.

I force pushed the changes, however the Azure Pipelines builds seem to be failing with an [unrelated error](https://dev.azure.com/cake-build/Cake/_build/results?buildId=4626&view=logs&jobId=00af5505-930d-50dd-3844-b167e8b4bb13&taskId=0fadbdea-3d1d-5635-586e-134b382e6b33&lineStart=1380&lineEnd=1381&colStart=1&colEnd=1) in the integration tests:
```
System.PlatformNotSupportedException: Operation is not supported on this platform.
   at Microsoft.CSharp.CSharpCodeGenerator.FromFileBatch(CompilerParameters options, String[] fileNames)
   at Microsoft.CSharp.CSharpCodeGenerator.System.CodeDom.Compiler.ICodeCompiler.CompileAssemblyFromDomBatch(CompilerParameters options, CodeCompileUnit[] ea)
   at Mono.TextTemplating.TemplatingEngine.CompileCode(IEnumerable`1 references, TemplateSettings settings, CodeCompileUnit ccu) in /Users/mikayla/code/t4/Mono.TextTemplating/Mono.TextTemplating/TemplatingEngine.cs:line 271
   at Mono.TextTemplating.TemplatingEngine.CompileTemplateInternal(ParsedTemplate pt, String content, ITextTemplatingEngineHost host) in /Users/mikayla/code/t4/Mono.TextTemplating/Mono.TextTemplating/TemplatingEngine.cs:line 175
   at Mono.TextTemplating.TemplatingEngine.CompileTemplate(ParsedTemplate pt, String content, ITextTemplatingEngineHost host) in /Users/mikayla/code/t4/Mono.TextTemplating/Mono.TextTemplating/TemplatingEngine.cs:line 149
   at Mono.TextTemplating.ToolTemplateGenerator.ProcessTemplate(ParsedTemplate pt, String inputFile, String inputContent, String& outputFile) in /Users/mikayla/code/t4/dotnet-t4/ToolTemplateGenerator.cs:line 59
   at Mono.TextTemplating.TextTransform.MainInternal(String[] args) in /Users/mikayla/code/t4/dotnet-t4/TextTransform.cs:line 194
   at Mono.TextTemplating.TextTransform.Main(String[] args) in /Users/mikayla/code/t4/dotnet-t4/TextTransform.cs:line 42
An error occurred when executing task 'Cake.Common.Tools.TextTransform.TextTransformAliases.TransformTemplate.Properties'.
```

patriksvensson commented:
@gitfool The T4 issue have been resolved, but I need to investigate the failing tests a bit. I will try to investigate this as soon as I have some spare time.

gitfool commented:
@patriksvensson re the unit tests that are now failing on *nix, I changed them from `WindowsTheory` to `Theory` because I thought they should now work on all platforms after I changed `ScriptAnalyzerFixture` ctor to create a Unix environment when specified.

I could change the attribute back but I'm still curious why they wouldn't work as the file system should be virtualized for the test.

gitfool commented:
I would guess the problem is coming from:
- [Cake.Testing/FakeEnvironment.cs#L79](https://github.com/cake-build/cake/blob/develop/src/Cake.Testing/FakeEnvironment.cs#L79)
  - `environment.WorkingDirectory = new DirectoryPath("C:/Working");`
- [Cake.Core/IO/Path.cs#L98](https://github.com/cake-build/cake/blob/develop/src/Cake.Core/IO/Path.cs#L98)
  - `IsRelative = !PathHelper.IsPathRooted(FullPath);`
- [Cake.Core/IO/PathHelper.cs#L13](https://github.com/cake-build/cake/blob/develop/src/Cake.Core/IO/PathHelper.cs#L13)
  - `_isWindows = EnvironmentHelper.GetPlatformFamily() == PlatformFamily.Windows;`

i.e. `PathHelper` does not honor the fake environment.

gitfool commented:
@patriksvensson I'm keen to get this PR into the next release, and since the issue with the tests is due to a pre-existing deficiency in the fake test environment, I simply reverted the Windows tests to use `WindowsTheory`, as before.

gitfool commented:
Rebased and squashed the last revert commit for a cleaner merge.

gitfool commented:
@devlead I added an integration test that adds a task dependency if loaded. One of the tasks throws an exception, but this does not happen since the globber [pattern](https://github.com/cake-build/cake/pull/2405/commits/b1b59c5b6415428a9c15c5e55a6aee3b56403a24#diff-bcaa202342012630dbfec8f59e0cca0eR15) excludes it.

Note: I also tweaked the check for `.cake` file extension to be earlier so that a warning is logged if no scripts are found _after_ filtering out any files without such an extension. I added some unit tests for this case too.

devlead commented:
@gitfool your changes have been merged, thanks for your contribution 👍
